### PR TITLE
promql: Add support for .Select calls concurrency. #6878

### PR DESCRIPTION
**Pull request by liguozhong**
promql: Add support for .Select calls concurrency. #6878

### CODE REVIEW

**Comment by brian-brazil** [comment]
This is still going to create goroutines when there's only the local tsdb to talk to. We want to talk remotely concurrently, but not locally.

**Comment by roidelapluie** [comment]
> This is still going to create goroutines when there's only the local tsdb to talk to. We want to talk remotely concurrently, but not locally.

Like #6770?

**Comment by liguozhong** [comment]
> This is still going to create goroutines when there's only the local tsdb to talk to. We want to talk remotely concurrently, but not locally.

done.

image

------- [9090 dev debug info]--------
image
image

------- [9010 dev debug info]--------
image
image

File: promql/engine.go [hunk 1]
```diff
@@ -651,57 +654,132 @@ func (ng *Engine) populateSeries(ctx context.Context, querier storage.Querier, s
...
+		}
+		parser.Inspect(s.Expr, func(node parser.Node, path []parser.Node) error {
+			switch node.(type) {
+			case *parser.VectorSelector:
+				RemoteSelectCalls++
+			}
+			return nil
+		})
+	}
+	if RemoteSelectCalls < 2 {
```
Comment (slrtbtfs) [inline promql/engine.go:670]: Would it significantly affect performance to move that `if` inside the `switch` and  thus avoid code duplication?

Comment (liguozhong) [reply promql/engine.go:667]: [outside]
image
[inside]
image

File: promql/engine.go [hunk 2]
```diff
@@ -651,57 +653,133 @@ func (ng *Engine) populateSeries(ctx context.Context, querier storage.Querier, s
 		warnings  storage.Warnings
 		err       error
 	)
+	RemoteSelectCalls := 0
+	if querier.Remotely() {
```
Comment (brian-brazil) [inline promql/engine.go:657]: PromQL shouldn't care about how the querier is implemented, it should always be the same logic.

Comment (liguozhong) [reply promql/engine.go]: thanks for your review.
I move "remotely" to the engine. Is that ok? Only three files need to be changed.
image
image
image

Comment (brian-brazil) [reply promql/engine.go]: That sort of check shouldn't appear anywhere in the promql directory, any special handling of local vs remote should be in fanout.

Comment (liguozhong) [reply promql/engine.go]: Thanks, I will think about how to implement this feature in "fanout.go"

**Comment by krasi-georgiev** [comment]
> This is still going to create goroutines when there's only the local tsdb to talk to. We want to talk remotely concurrently, but not locally.

Why not locally as well?

**Comment by brian-brazil** [comment]
> Why not locally as well?

It'd make queries more expensive overall, and would also mean that there was no longer any limit to how many cores PromQL could use - neither of which is desirable.

File: promql/engine.go (hunk 2 above)
Comment (liguozhong) [reply promql/engine.go]: I have moved the remote related logic to the fanout directory.
done . please review again ,thanks.
image
debug info 9010 is remotely result, 9090 is locally result
image

File: promql/engine.go [hunk 3]
```diff
@@ -689,19 +687,40 @@ func (ng *Engine) populateSeries(ctx context.Context, querier storage.Querier, s
...
+			sp := &storage.SelectParam{SortSeries: false, Hints: hints, Matchers: n.LabelMatchers}
+			selectParams = append(selectParams, sp)
+			selectCalls[sp] = n
 		case *parser.MatrixSelector:
 			evalRange = n.Range
 		}
 		return nil
 	})
+	switch q := querier.(type) {
+	case storage.BatchQuerier:
```
Comment (brian-brazil) [inline promql/engine.go:699]: You're still special casing here, the engine should always be doing the same thing no matter what storage it is being passed.

**Comment by krasi-georgiev** [comment]
> It'd make queries more expensive overall, and would also mean that there was no longer any limit to how many cores PromQL could use - neither of which is desirable.

yes this it the tradeoff, but some use cases might have enough resources and would want that faster processing.

> and would also mean that there was no longer any limit to how many cores PromQL could use - neither of which is desirable.

we could limit the goroutines to the number of available cores, no?

**Comment by roidelapluie** [comment]
> > It'd make queries more expensive overall, and would also mean that there was no longer any limit to how many cores PromQL could use - neither of which is desirable.
> 
> yes this it the tradeoff, but some use cases might have enough resources and would want that faster processing.
> 
> > and would also mean that there was no longer any limit to how many cores PromQL could use - neither of which is desirable.
> 
> we could limit the goroutines to the number of available cores, no?

not really as one core should be reserved for compaction

**Comment by krasi-georgiev** [comment]
yes sure, my point was that we could still use concurrency but also limit the total goroutines.

**Comment by bwplotka** [comment]
Ok, I am confused here. 

> This is still going to create goroutines when there's only the local tsdb to talk to. We want to talk remotely concurrently, but not locally.

What exactly this means? PromQL can have 100 Select to make and it will do it sequentially, no matter what storage is right now (: Let's say someone wants to scale read part of Prometheus (or in fact any user of `promql` package). So are we ok to add concurrency here or not?

File: promql/engine.go (hunk 3 above)
Comment (liguozhong) [reply promql/engine.go:699]: I see. I'll fix it right away

**Comment by brian-brazil** [comment]
> So are we ok to add concurrency here or not?

We want concurrency, but not for the tsdb so as to limit CPU usage to one core so that it's all predictable. This is in line with the original issue, which is about remote read queries.

**Comment by roidelapluie** [comment]
> > So are we ok to add concurrency here or not?
> 
> We want concurrency, but not for the tsdb so as to limit CPU usage to one core so that it's all predictable. This is in line with the original issue, which is about remote read queries.

When we run 10 queries in parallel, there is no select happening in parallel?

**Comment by slrtbtfs** [comment]
> we could limit the goroutines to the number of available cores, no?

The select goroutines will spend a lot of time waiting on I/O, so even if we have as many goroutines as cores we might not have full CPU load.

Also, schedulers exist, so even in case of full CPU load other tasks would still get some attention.

**Comment by brian-brazil** [comment]
> The select goroutines will spend a lot of time waiting on I/O,

It's mostly CPU based, Select only touches the indexes which will tend to be in RAM.

> Also, schedulers exist, so even in case of full CPU load other tasks would still get some attention.

Allowing PromQL queries to use unbounded CPU is a bad idea, that'll affect ingestion and other things on the machine. Overload is overload, schedulers don't change that.

> When we run 10 queries in parallel, there is no select happening in parallel?

There is, limited to no more than 10 cores.

**Comment by slrtbtfs** [comment]
Wouldn't it work to limit the total number of concurrent Select Calls in the storage implementation?

In that case local storage could always do only one select call at once and remote storage can have a higher limit.

**Comment by roidelapluie** [comment]
> Wouldn't it work to limit the total number of concurrent Select Calls in the storage implementation?
> 
> In that case local storage could always do only one select call at once and remote storage can have a higher limit.

That would make little queries a lot longer if they have to wait for long queries to run

**Comment by brian-brazil** [comment]
> In that case local storage could always do only one select call at once and remote storage can have a higher limit.

That's basically what I'm saying.

> That would make little queries a lot longer if they have to wait for long queries to run

That's already the case today, and it hasn't caused any issues I'm aware of.

File: promql/engine.go (hunk 3 above)
Comment (liguozhong) [reply promql/engine.go]: done.

code /promql/
image
code /fanout/
image

**Comment by roidelapluie** [comment]
What I noted on #6878 is that we would need a design document and I have yet to see it.

**Comment by roidelapluie** [comment]
> What I noted on #6878 is that we would need a design document and I have yet to see it.

in https://github.com/prometheus/prometheus/pull/6887 *

**Comment by brancz** [comment]
@roidelapluie where did you say that? I went over there and tried to search for it, but couldn't find it.

I personally don't think we need a design for this. I'm not saying we shouldn't discuss the strategy, but a full blown design doc seems exaggerated.

**Comment by roidelapluie** [comment]
> @roidelapluie where did you say that? I went over there and tried to search for it, but couldn't find it.
> 
> I personally don't think we need a design for this. I'm not saying we shouldn't discuss the strategy, but a full blown design doc seems exaggerated.

I did not say that but the previous PR tended to show that we should not run queries in parallel against local storage

**Comment by brian-brazil** [comment]
Yeah, this requires a bit of thought and upfront design but nothing that can't be done in the context of an issue/PR.

**Comment by brancz** [comment]
@roidelapluie sorry misunderstood your comment then. Seems like we’re in agreement :)

**Comment by liguozhong** [comment]
The idea of this PR is to first convert select() to batch query by adding a func of selects(). On the basis of batch query, add support for .Select calls concurrency. 
The screenshot is the flow chart of this PR.

image

File: pkg/gate/gate_test.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+	maxGoRoutine := 3
+	gate := New(maxGoRoutine)
+	callCount := 10
+	overloadChan := make(chan int, callCount-maxGoRoutine)
+	var wg sync.WaitGroup
+
+	for i := 0; i < callCount; i++ {
+		go func() {
+			wg.Add(1)
```
Comment (slrtbtfs) [inline pkg/gate/gate_test.go:34]: The `wg.Add(1)` must happen before launching the goroutine, otherwise this is a race condition.

Comment (liguozhong) [reply pkg/gate/gate_test.go]: thanks，done

**Comment by bwplotka** [comment]
Hey, @liguozhong are you available to continue this PR? I and @kakkoyun might have internally more time to move this forward this sprint if you need some help. I think we still need some approved designs for this. I might think we need to elaborate more than just couple of screenshots. I like Google Docs because we can comment easily on each stuff, but we could do this here or on https://github.com/prometheus/prometheus/issues/6878. Let's work together on some design that would do if you are ok with that. (:

**Comment by liguozhong** [comment]
> Hey, @liguozhong are you available to continue this PR? I and @kakkoyun might have internally more time to move this forward this sprint if you need some help. I think we still need some approved designs for this. I might think we need to elaborate more than just couple of screenshots. I like Google Docs because we can comment easily on each stuff, but we could do this here or on #6878. Let's work together on some design that would do if you are ok with that. (:

thanks . I have no experience with Google Docs, and I agree to let other contributors complete this PR. Looking forward to seeing this feature in 2.18.0 release.
I will continue to pay attention to this issue, our production environment also encountered this problem.
My experience is just to write code, if there are any work items that I can accomplish, I am happy to complete.

**Comment by kakkoyun** [comment]
@liguozhong Awesome, we will try to move this forward then with some design first. Make sure we have a good understanding of the issue and a consensus on implementation details.

**Comment by liguozhong** [comment]
> @liguozhong Awesome, we will try to move this forward then with some design first. Make sure we have a good understanding of the issue and a consensus on implementation details.

Looking forward to completion soon

**Comment by bwplotka** [comment]
Let's close this for now until we have agreed design in place. :hugs:

**Pull request closed without merging**